pipenv/core.py

```diff
@@ -1574,6 +1548,166 @@ def pip_install(
+    if not_hashed_requirements:
+        files.append(not_hashed_requirements)
+    for file in files:
+        pip_command = [
+            project_python(project, system=allow_global),
+            _get_runnable_pip(),
+            "install",
+        ]
+        pip_args = get_pip_args(
+            project,
+            pre=pre,
+            verbose=False,  # When True, the subprocess fails to recognize the EOF when reading stdout.
+            upgrade=True,
+            selective_upgrade=selective_upgrade,
+            no_use_pep517=not use_pep517,
+            no_deps=no_deps,
+        )
+        pip_command.extend(pip_args)
+        pip_command.extend(["-r", normalize_path(file.name)])
+        if dev and use_constraint:
+            default_constraints = get_constraints_from_deps(project.packages)
+            constraint_filename = prepare_constraint_file(
+                default_constraints,
+                directory=requirements_dir,
+                sources=None,
+                pip_args=None,
+            )
+            pip_command.extend(["-c", normalize_path(constraint_filename)])
+        if project.s.is_verbose():
+            msg = f"Install Phase: {'Requirements with Hashes' if file == hashed_requirements else 'Requirements without Hashes'}"
+            click.echo(
+                click.style(msg, bold=True),
+                err=True,
+            )
+            for requirement in (
+                hashed_deps if file == hashed_requirements else not_hashed_deps
+            ):
+                click.echo(
+                    click.style(
+                        f"Preparing Installation of {requirement.name!r}", bold=True
+                    ),
+                    err=True,
+                )
+            click.echo(f"$ {cmd_list_to_shell(pip_command)}", err=True)
+        cache_dir = Path(project.s.PIPENV_CACHE_DIR)
+        default_exists_action = "w"
+        if selective_upgrade:
+            default_exists_action = "i"
+        exists_action = project.s.PIP_EXISTS_ACTION or default_exists_action
+        pip_config = {
+            "PIP_CACHE_DIR": cache_dir.as_posix(),
+            "PIP_WHEEL_DIR": cache_dir.joinpath("wheels").as_posix(),
+            "PIP_DESTINATION_DIR": cache_dir.joinpath("pkgs").as_posix(),
+            "PIP_EXISTS_ACTION": exists_action,
+            "PATH": os.environ.get("PATH"),
+        }
+        if src_dir:
+            if project.s.is_verbose():
+                click.echo(f"Using source directory: {src_dir!r}", err=True)
+            pip_config.update({"PIP_SRC": src_dir})
+        c = subprocess_run(pip_command, block=False, capture_output=True, env=pip_config)
+        if file == hashed_requirements:
+            c.deps = hashed_deps
+        else:
+            c.deps = not_hashed_deps
+        c.env = pip_config
+        cmds.append(c)
+        while True:
+            line = c.stdout.readline()
+            if line == "":
```

Comment:
This was really problematic when the pip logger was overriden or verbose was passed to pip, we never recieved the EOF and no matter what the process.poll() or return_code are never available, as if the process doesn't get cleaned up until the cleanup routine.   Though the cleanup routine uses `p.communicate()` which doesn't get the output until the process returns at the end so there was no indicator of progress without this logic here.   This gives real time progress logic when verbose is passed.